### DISCUSSION TITLE
Storing metadata in the database

### DISCUSSION THREAD

**Discussion by nielstron** [CardanoSolutions/kupo#125] [Q&A]
### What is your idea? Provide a use case.
When querying the transaction history of DEXs, it is interesting to see what metadata was provided with each transaction (i.e. to distinguish order placed through the main interface, trading bots and DEX Aggregators).
This requires quick access to transaction metadata.

### Why is it a good idea?

By storing the metadata in a local table in kupo, access time can be reduced to logarithmic time. By pruning away metadata unrelated to the stored transactions, this wouldn't affect the size of the database too much either.

### What is the current alternative and why is it not good enough?

Currently kupo offers an option to access metadata for a transaction. Access time is easily on the order of 1-2s though since the metadata is not stored in the local database, but fetched from the node. This makes analysis of DEX transactions (easily exceeding 1M unique transactions) infeasible.

Another alternative is fetching data from blockfrost or seperate instance, which however seperates sources and may lead to problems when accessing "fresh" data on chain (one instance having picked up a transaction with the other instance missing it)

**Comment by KtorZ** [comment]
Current access time for metadata is already logarithmic. 

> Access time is easily on the order of 1-2s

My own experience and benchmarks have been more in the range of 10-20ms. Do you have any particular reproducible case / setup where you observe such large delays ?

**Comment by nielstron** [comment]
I will try to set up something reproducible! If you want to play around with it in the meantime, this example comes from the MinSwap orderbook contract, transactions close to launch of the platform on mainnet.

**Reply by nielstron** [reply]
Just to update this: I have tried the following query on mainnet

```
time curl <kupo>/metadata/54833279
```

And could not reproduce access times around 1-2s. However I found that tail latency might be quite high and appears to generally depend on the load on the cardano-node.

Server without load on node: all requests around 0.02s = 20ms
Server with medium load on the node (i.e. parallel cardano-cli, Ogmios and kupo with load running): *most* requests around 20ms but maximum was 300ms.

I have recently drastically reduced the load on the node (using kupo) so I am assuming that the measurement fo 1-2s was done when there was more load (i.e. permanent calls to cardano-cli fetching large parts of the UTxO).

Ideally I would wish for kupo to enable taking load off the node and since metadata is already present when fetching blocks for syncing, storing the metadata in kupo would reduce the load for the node. I acknowledge though that this might also add more load on kupo/the supposedly minimal sqlite db so I understand if the metadata is not added into the db.

**Comment by biofreeze** [comment]
I have a use case for this as well. Storing metadata in the database would be an awesome feature. My use case is for a simple API that can get the assets within a user's wallet for example. Right now Kupo is great because I can get how much Cardano (and other tokens) and # of NFT's a user has, but obviously the output looks like `"f96584c4fcd13cd1702c9be683400072dd1aac853431c99037a3ab1e.43616c69":1` when returned from the `/matches` endpoint. I'd love to be able to specify a flag where metadata for each asset is included in the result (which would give me access to the NFT name, image, etc).

I tried using the `/metadata/[slot]` endpoint, but it's just not as fast as it would need to be unfortunately.

**Reply by KtorZ** [reply]
How fast would it need to be to be satisfactory?

**Reply by biofreeze** [reply]
Well fast enough for serving client facing queries! Like I shared in my post above, I would love to use Kupo to get information about user's assets; which obviously it can already, but it's just missing that other key information regarding each asset (i.e - the metadata (name of the asset, image, etc). If metadata could be included in the matches result (turned on my an option/flag) this would be even better.

I think if we had quick access to metadata Kupo could truly replace the need for other options like Oura, and Cardano DB Sync (for simple things like i'm trying to do)

**Reply by biofreeze** [reply]
Also being able to lookup metadata in the database given a specific `policy` or `policy + asset_name` would be sweet too.

**Reply by biofreeze** [reply]
Right now, my solution is just to use Maestro and query the `https://mainnet.gomaestro-api.org/v1/policy/[policy]/assets` endpoint to get metadata for each policy that my app allows, and store it all in another local SQLite db, alongside the Kupo db. 

So when I need to get a user's assets, I'll query the `/matches` endpoint, get all the assets (which will be like `"f96584c4fcd13cd1702c9be683400072dd1aac853431c99037a3ab1e.43616c69":1` and then query for the metadata in the other SQLite db that holds all the metadata, by either `policy` or `policy + asset`)

But having Kupo store metadata in the db would eliminate the need to use the Maestro API to query metadata essentially 😃

**Reply by KtorZ** [reply]
But I am slightly confused, you're saying it is "too slow" whereas the various benchmarks I have show that metadata can be retrieved within a dozen of milliseconds.

I don't see how querying another remote service (eg Maestro) can be significantly faster here? 

For example, in the starter kit for Demeter, the example number 3 fetches all metadata associated with Hosky holders of the past 24h. The example (https://github.com/CardanoSolutions/kupo-js-starter-kit/tree/main/examples/3-find-hosky-hodlers) typically fetches ten of thousands of results and their metadata within seconds. 

Truth is, metadata are already stored in a database, but the node's one. Kupo has a direct access to them in a logarithmic time complexity. Duplicating the data in a new database, while it would possibly provide new index keys I concede, wouldn't lead to much asymptotic improvements. 

One idea mentioned on another discussion could be to have Kupo do some "deep fetching" of metadata and fetch them alongside the query. This wouldn't be significantly faster but at least it would be more convenient (as clients would get everything in a single request).

**Reply by biofreeze** [reply]
I take a performance first approach to everything, so I'd prefer if the data was just there ready for querying, instead of having to go and fetch it, is mostly why I suggested storing it in the database. But doing the "deep fetching" approach you suggested could also work, as long as it still works well under medium/heavy load 👍

**Reply by KtorZ** [reply]
But the data IS there. That's what I've been kindly explaining ☺️. Performance matters, yes. But there's no point discussing performance without measurements. 

Right now, the system can accommodate serving metadata at a rather decent rate (~hundred of metadata per second). If you need something higher than that, I'd be curious to hear about the use case. Kupo may simply not be the best fit for you. 

I am also keen on exploring options which remains reasonable, yet for now your claim seems more of a gut feeling than a measured observation?
